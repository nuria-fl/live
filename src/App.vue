<template>
  <main>
    <section v-if="!hasStarted" class="Splash">
      <h1>Live</h1>
      <p>A game about survival</p>
      <p>What's your name, survivor?</p>
      <form @submit.prevent="start">
        <input
          v-model="startUsername"
          type="text"
          required
          class="Splash__input"
        />
        <button type="submit" class="Btn">
          New Game
        </button>
      </form>
    </section>
    <main-component v-else @newGame="start" />
    <footer class="Footer">
      <game-status-button v-if="hasStarted" />
      Bug report via
      <a href="https://github.com/nuria-fl/live">Github</a>
      or
      <a href="https://twitter.com/pincfloit">Twitter</a>
      · ☕ <a href="https://ko-fi.com/G2G5AV2X">Buy Me a Coffee</a>
    </footer>
  </main>
</template>

<script>
import { mapState, mapMutations } from 'vuex'
import Main from '@/views/Main'
import GameStatusButton from '@/components/GameStatusButton'
import tracking from '@/utils/tracking'

export default {
  components: {
    MainComponent: Main,
    GameStatusButton
  },
  data() {
    return {
      hasStarted: false,
      startUsername: ''
    }
  },
  computed: {
    ...mapState(['username'])
  },
  created() {
    this.initUsername()
    this.startUsername = this.username
  },
  methods: {
    ...mapMutations(['initUsername', 'setUsername']),
    start() {
      tracking.trackEvent('game', 'start')
      this.setUsername(this.startUsername)
      this.hasStarted = true
    }
  }
}
</script>

<style lang="scss">
body {
  margin: 0;
  font-family: 'Fira Mono', monospace;
  font-size: 0.95em;
  overscroll-behavior: contain;
}
li {
  margin: 1em 0;
}
.Splash {
  padding: 3em 0 4em;
  text-align: center;
  &__input {
    margin-right: 0.5rem;
  }
}
.Footer {
  width: 100%;
  margin-top: 2em;
  padding: 1em;
  background: rgba(255, 255, 255, 0.8);
  @media screen and (min-width: 680px) {
    position: fixed;
    bottom: 0;
    left: 0;
  }
}
</style>
