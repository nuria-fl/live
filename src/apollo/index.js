import ApolloClient from 'apollo-boost'

export const apolloClient = new ApolloClient({
  uri: process.env.VUE_APP_API_URL
})
