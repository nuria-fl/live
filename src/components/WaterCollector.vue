<template>
  <article class="Item">
    <h4>Water collector</h4>
    <template v-if="!hasWaterCollector">
      <p>
        Items needed:
        {{ itemsNeeded }}
        <br />
        Build a water collector to get rain water
      </p>
      <div class="Item__actions">
        <button
          :disabled="!item.isCraftable || disabled"
          type="button"
          class="Btn"
          @click="buildWaterCollector()"
        >
          Craft
        </button>
      </div>
    </template>
    <template v-else-if="isCollecting">
      Uses remaining: {{ usesRemaining - 1 }}<br />
      <span class="progress">
        Collecting water
      </span>
    </template>
    <template v-else>
      Uses remaining: {{ usesRemaining - 1 }}<br />
      <button :disabled="disabled" class="Btn" @click="drinkWater">
        Drink
      </button>
      <button :disabled="disabled" class="Btn" @click="collectWater">
        Collect
      </button>
    </template>
  </article>
</template>

<script>
import { mapState, mapMutations, mapActions } from 'vuex'
import eventBus from '@/utils/eventBus'
import items from '@/utils/items'

export default {
  props: {
    item: {
      type: Object,
      required: true
    }
  },
  data() {
    return {
      hasWaterCollector: false,
      isCollecting: false,
      isReady: false,
      usesRemaining: 3,
      currentTime: 0,
      waterTimeout: null
    }
  },
  computed: {
    ...mapState(['disabled', 'inventory', 'existingItems']),
    itemsNeeded() {
      return this.item.itemsNeeded.map(items.getName).join(', ')
    }
  },
  mounted() {
    eventBus.$on('gameStatusChange', this.handleGameStatusChange)
  },
  methods: {
    ...mapMutations(['increase', 'addInventory']),
    ...mapActions(['removeItemsById']),
    buildWaterCollector() {
      this.removeItemsById(this.item.itemsNeeded)
      this.usesRemaining = 3
      this.hasWaterCollector = true
      this.startCollecting()
    },
    startCollecting() {
      this.isCollecting = true
      this.startCollectingLoop()
    },
    startCollectingLoop() {
      this.waterTimeout = setTimeout(() => {
        this.currentTime++
        if (this.currentTime === 60) {
          this.isCollecting = false
          this.currentTime = 0
        } else {
          this.startCollectingLoop()
        }
      }, 1000)
    },
    drinkWater() {
      this.increase({
        stat: 'water',
        amount: 20
      })
      this.finishUse()
    },
    collectWater() {
      const hasBottle = this.inventory.find((item) => item.id === 'bottle')
      if (hasBottle) {
        const water = this.existingItems.find(
          (item) => item.id === 'water-clean'
        )
        this.removeItemsById(['bottle'])
        this.addInventory(water)
        this.finishUse()
      } else {
        eventBus.$emit('showModal', {
          body: 'You need an empty bottle to collect water'
        })
      }
    },
    finishUse() {
      this.usesRemaining--

      if (this.usesRemaining === 0) {
        this.hasWaterCollector = false
        eventBus.$emit('showNotification', {
          text: 'Water collector has broken'
        })
      } else {
        this.startCollecting()
      }
    },
    pauseWaterCollector() {
      clearTimeout(this.waterTimeout)
    },
    handleGameStatusChange(isPaused) {
      if (this.hasWaterCollector) {
        if (isPaused) {
          this.pauseWaterCollector()
        } else {
          this.startCollectingLoop()
        }
      }
    }
  }
}
</script>
