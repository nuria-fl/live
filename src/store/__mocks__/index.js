import Vue from 'vue';
import Vuex from 'vuex';
import items from '../../data/items'

Vue.use(Vuex);

export const getters = {
  isInventoryFull: jest.fn().mockReturnValue(false),
  slotsInInventoryLeft: jest.fn().mockReturnValue(8),
  craftableItems: jest.fn().mockReturnValue()
};

export const mutations = {
  pauseGame: jest.fn(),
  playGame: jest.fn(),
  changePage: jest.fn(),
  increaseDayCount: jest.fn(),
  decrease: jest.fn(),
  increase: jest.fn(),
  addInventory: jest.fn(),
  removeInventory: jest.fn(),
  disable: jest.fn(),
  enable: jest.fn(),
  enableFire: jest.fn(),
  disableFire: jest.fn(),
  getSick: jest.fn(),
  getCured: jest.fn()
};

export const actions = {
  initInventory: jest.fn(),
  decreaseAsync: jest.fn(),
  increaseAsync: jest.fn(),
  scavenge: jest.fn(),
  hunt: jest.fn(),
  removeItemsById: jest.fn()
};

export const state = {
  gameOver: false,
  causeOfDeath: null,
  paused: false,
  disabled: false,
  stats: {
    health: 100,
    water: 100,
    food: 100,
    sleep: 100
  },
  existingItems: items,
  inventory: [],
  hasFire: false,
  isSick: false,
  daysSurvived: 0,
  currentPage: 'home'
};

// eslint-disable-next-line no-underscore-dangle
export function __createMocks(custom = { getters: {}, mutations: {}, actions: {}, state: {} }) {
  const mockGetters = Object.assign({}, getters, custom.getters);
  const mockMutations = Object.assign({}, mutations, custom.mutations);
  const mockActions = Object.assign({}, actions, custom.actions);
  const mockState = Object.assign({}, state, custom.state);

  return {
    getters: mockGetters,
    mutations: mockMutations,
    actions: mockActions,
    state: mockState,
    store: new Vuex.Store({
      getters: mockGetters,
      mutations: mockMutations,
      actions: mockActions,
      state: mockState,
    }),
  };
}

export const store = __createMocks().store;
