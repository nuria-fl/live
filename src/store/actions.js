import utils from '@/utils'
import eventBus from '@/utils/eventBus'

export default {
  initInventory({ state, commit }) {
    const scavengeableItems = state.existingItems.filter(
      (item) => item.action === 'scavenge'
    )
    const startingItems = utils.randomizeItems(scavengeableItems, 2)
    startingItems.forEach((item) => commit('addInventory', item))
  },
  decreaseAsync({ state, commit }, { stat, amount, time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('enable')
          commit('decrease', { stat, amount })
          resolve()
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  increaseAsync({ state, commit }, { stat, amount, time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('enable')
          commit('increase', { stat, amount })
          resolve()
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  scavenge({ state, getters, commit }, { time }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          const scavengeableItems = state.existingItems.filter(
            (item) => item.action === 'scavenge'
          )

          const numberOfItems =
            getters.slotsInInventoryLeft > 3 ? 3 : getters.slotsInInventoryLeft

          const newItems = utils.randomizeItems(
            scavengeableItems,
            numberOfItems
          )

          newItems.forEach((item) => {
            commit('addInventory', item)
          })

          commit('decrease', { stat: 'energy', amount: 5 })
          commit('decrease', { stat: 'water', amount: 5 })
          commit('decrease', { stat: 'food', amount: 3 })
          commit('enable')

          resolve(newItems)
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  hunt({ state, commit, dispatch }, { time, weapon }) {
    return new Promise((resolve, reject) => {
      commit('disable')
      if (!state.gameOver) {
        setTimeout(function() {
          commit('decrease', { stat: 'energy', amount: 10 })
          commit('decrease', { stat: 'water', amount: 10 })
          commit('decrease', { stat: 'food', amount: 6 })
          commit('enable')

          const isSuccesful = utils.calculateProbability(8)

          if (isSuccesful) {
            dispatch('handleItemDegradation', weapon)
            const huntableItems = state.existingItems.filter(
              (item) => item.action === 'hunt'
            )

            const newItems = utils.randomizeItems(huntableItems, 1)

            newItems.forEach((item) => {
              commit('addInventory', item)
            })

            resolve(newItems)
          } else {
            resolve(false)
          }
        }, time)
      } else {
        reject(new Error('game is over'))
      }
    })
  },
  handleItemDegradation({ commit }, item) {
    if (item.usesUntilBreakdown > 1) {
      commit('degradeItem', item.uid)
    } else {
      commit('removeInventory', item.uid)
      eventBus.$emit('showNotification', {
        text: `${item.name} has broken`
      })
    }
  },
  removeItemsById({ state, commit }, items) {
    const findItemByName = (itemName) => {
      return state.inventory.find((item) => item.id === itemName)
    }

    items.forEach((item) => {
      const itemToRemove = findItemByName(item)
      commit('removeInventory', itemToRemove.uid)
    })
  }
}
