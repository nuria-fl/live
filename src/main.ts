import "./assets/styles/index.scss";

import { createApp } from "vue";

import App from "./App.vue";
import { createAppStore } from "./store";

const store = createAppStore();

const app = createApp(App);

app.use(store);

app.mount("#app");
