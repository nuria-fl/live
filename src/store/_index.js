import { createStore } from "vuex";

import actions from "./_actions";
import mutations from "./_mutations";
import { getters, state } from "./_state";

const store = createStore({
	state,
	getters,
	mutations,
	actions,
});

export default store;
