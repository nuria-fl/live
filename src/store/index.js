import { createStore } from 'vuex'
import { state, getters } from './state'
import mutations from './mutations'
import actions from './actions'

const store = createStore({
  state,
  getters,
  mutations,
  actions
})

export default store
