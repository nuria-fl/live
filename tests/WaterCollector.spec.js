import Vuex from 'vuex'
import { shallow, createLocalVue } from 'vue-test-utils'
import { __createMocks as createStoreMocks } from '../src/store'

import WaterCollector from '../src/components/WaterCollector'

// Tell Jest to use the mock implementation of the store
jest.mock('../src/store')

const localVue = createLocalVue()

localVue.use(Vuex)

describe('WaterCollector', () => {
  let storeMocks
  let wrapper

  beforeEach(() => {
    // Create a fresh store and wrapper instance for every test case.
    storeMocks = createStoreMocks()
    wrapper = shallow(WaterCollector, {
      store: storeMocks.store,
      localVue,
      propsData: {
        item: {
          id: 'water-collector',
          type: 'camp',
          name: 'Water collector',
          items: ['plastic', 'rope', 'bottle'],
          isCraftable: true
        }
      }
    })
  })

  test('It should match snapshot', () => {
    wrapper.find('.Btn').trigger('click')
    expect(wrapper.vm.$el).toMatchSnapshot()
  })

  test('It should call removeItemsById when crafted', () => {
    wrapper.find('.Item__actions .Btn').trigger('click')
    expect(storeMocks.actions.removeItemsById).toBeCalled()
  })

  test('It should set hasWaterCollector to true when crafted', () => {
    wrapper.find('.Item__actions .Btn').trigger('click')
    expect(wrapper.vm.hasWaterCollector).toBe(true)
  })

  test('It should set isCollecting to true when crafted', () => {
    wrapper.find('.Item__actions .Btn').trigger('click')
    expect(wrapper.vm.isCollecting).toBe(true)
  })
})
