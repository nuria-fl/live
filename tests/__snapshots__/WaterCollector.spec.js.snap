// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WaterCollector It should match snapshot 1`] = `
<article
  class="Item"
>
  <h4>
    Water collector
  </h4>
   
    Uses remaining: 2
  <br />
   
  <span
    class="progress"
  >
    
      Collecting water
    
  </span>
</article>
`;
